### PR TITLE
Feature plugins

### PR DESCRIPTION
**Pull request by SteveGilvarry**
Bringing @nextime 's work back into the fold, thanks to brilliant efforts by @manupap1 thought it was time to bring this work across into zoneminder repo.

Known issues:
Autotools not building
SQL changes need to be reversed out of 1.28.100 and put into a temporary home until we are ready to merge.
Haven't had much time to polish it yet so don't expect the branch to work yet

### CODE REVIEW

**Comment by knnniggett** [comment]
Not sure I'd worry about autotools. Perhaps we could comment out the autotools stuff in the Travis config file for this branch.

File: db/zm_update-1.28.1.sql [hunk 1]
```diff
@@ -1 +1,21 @@
 ALTER TABLE Monitors MODIFY Device tinytext;
+
```
Comment (SteveGilvarry) [inline db/zm_update-1.28.1.sql]: Rip this out

File: configure.ac [hunk 2]
```diff
@@ -4,11 +4,18 @@
...
+AC_CONFIG_AUX_DIR([.])
 AM_INIT_AUTOMAKE
+LT_PREREQ([2.4.2])
+LT_INIT([dlopen])
+AC_CONFIG_MACRO_DIR([m4])
 AC_CONFIG_SRCDIR(src/zm.h)
 AC_CONFIG_HEADERS(config.h)
 AC_CONFIG_MACRO_DIR([m4])
 
+ZM_ENGINE_VERSION=29
```
Comment (kylejohnson) [inline configure.ac:16]: What is a ZM_ENGINE_VERSION?

**Comment by SteveGilvarry** [comment]
It is a plugin compatibility version. I envision that sometimes we get a change that breaks plugins interface points. At which time we increment engine and plugins have an engine version so they know to not load if they mismatch. On phone so linking hard, but check zm_plugin.cpp line 103 I think.
I did check most of this as I merged, but was last week so forget some details. Still spinning it up and refining what I bring in so keep the comments coming.

File: CMakeLists.txt [hunk 3]
```diff
@@ -230,6 +261,7 @@ find_package(ZLIB)
 if(ZLIB_FOUND)
 	set(HAVE_LIBZLIB 1)
 	list(APPEND ZM_BIN_LIBS "${ZLIB_LIBRARIES}")
+	set(LIBS "${LIBS} -lz")
```
Comment (SteveGilvarry) [inline CMakeLists.txt:270]: Need to confirm this 100% but @connortechnology was asking about these. I believe they are to let plugins know what libs are present. libzoneminder is related just need to put it all together.

**Comment by manupap1** [comment]
please note that manupap1 died suddenly the 19th of august 2015.

best regards,
Mélodie Papin (his sister).

**Comment by SteveGilvarry** [comment]
FYI Since this is a branch no real harm in the self merge and it has been hanging around for a while as a PR. Feel free to keep asking questions but I think I have answered all the ones asked so far.

**Comment by haolez** [comment]
Is this still being worked on? How can I help?

**Comment by nitrag** [comment]
I would very much like to see this happen. I think a generic plugin to send/pipe an image to a local script (user defined) would be awesome.

**Pull request merged**